compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -660,13 +678,15 @@ pub enum UpvarSubsts<'tcx> {
 }
 
 impl<'tcx> UpvarSubsts<'tcx> {
+    /// Returns an iterator over the list of types of captured paths by the closure/generator.
+    /// In case there was a type error in figuring out the types of the captured path, an
+    /// empty iterator is returned.
     #[inline]
     pub fn upvar_tys(self) -> impl Iterator<Item = Ty<'tcx>> + 'tcx {
-        let tupled_upvars_ty = match self {
-            UpvarSubsts::Closure(substs) => substs.as_closure().split().tupled_upvars_ty,
-            UpvarSubsts::Generator(substs) => substs.as_generator().split().tupled_upvars_ty,
-        };
-        tupled_upvars_ty.expect_ty().tuple_fields()
+        match self {
```

Comment:
```suggestion
        let tupled_tys = match self {
            UpvarSubsts::Closure(substs) => substs.as_closure().self.tupled_upvars_ty(),
            ...
        }
```

and then u can bring in a similar match block from the other upvar_ty impls